sdks/java/core/src/main/java/org/apache/beam/sdk/util/MovingFunction.java

```diff
@@ -69,25 +69,21 @@ public MovingFunction(
     numSamples = new int[n];
     Arrays.fill(numSamples, 0);
     currentMsSinceEpoch = -1;
-    currentIndex = -1;
+    currentIndex = 0;
   }
 
   /** Flush stale values. */
   private void flush(long nowMsSinceEpoch) {
     checkArgument(nowMsSinceEpoch >= 0, "Only positive timestamps supported");
-    if (currentIndex < 0) {
-      currentMsSinceEpoch = nowMsSinceEpoch - (nowMsSinceEpoch % sampleUpdateMs);
-      currentIndex = 0;
-    }
     checkArgument(nowMsSinceEpoch >= currentMsSinceEpoch, "Attempting to move backwards");
     int newBuckets =
         Math.min((int) ((nowMsSinceEpoch - currentMsSinceEpoch) / sampleUpdateMs), buckets.length);
+    currentMsSinceEpoch = nowMsSinceEpoch - (nowMsSinceEpoch % sampleUpdateMs);
     while (newBuckets > 0) {
       currentIndex = (currentIndex + 1) % buckets.length;
       buckets[currentIndex] = function.identity();
       numSamples[currentIndex] = 0;
       newBuckets--;
-      currentMsSinceEpoch += sampleUpdateMs;
```

Comment:
This was a bug? What was the effect? Looks like the buckets got cleared incorrectly?

Reply:
Yes, this was a bug. If a sample with `nowMsSinceEpoch > SAMPLE_PERIOD` was recorded, `currentMsSinceEpoch` would move forward no more than `SAMPLE_PERIOD` and stay behind `now`, next call to `get(now)` would incorrectly flush recorded value as it would be considered out of the current window (`nowMsSinceEpoch - currentMsSinceEpoch` determines which buckets to flush).

It had no effect if `get()` or `put()` were invoked every `SAMPLE_PERIOD` (which, i think, is mostly the case for streaming), then `currentMsSinceEpoch` would never stay behind.

Reply:
Thanks!